packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts

```diff
@@ -0,0 +1,857 @@
+     */
+    public async softCancelOrderAsync(order: Order | SignedOrder): Promise<CoordinatorServerCancellationResponse> {
+        assert.doesConformToSchema('order', order, schemas.orderSchema);
+        assert.isETHAddressHex('feeRecipientAddress', order.feeRecipientAddress);
+        assert.isSenderAddressAsync('makerAddress', order.makerAddress, this._web3Wrapper);
+
+        const data = this._transactionEncoder.cancelOrderTx(order);
+        const transaction = await this._generateSignedZeroExTransactionAsync(data, order.makerAddress);
+        const endpoint = await this._getServerEndpointOrThrowAsync(order.feeRecipientAddress);
+
+        const response = await this._executeServerRequestAsync(transaction, order.makerAddress, endpoint);
+        if (response.isError) {
+            throw new CoordinatorServerError(
+                CoordinatorServerErrorMsg.CancellationFailed,
+                [],
+                [],
+                [
+                    {
+                        ...response,
+                        orders: [order],
+                    },
+                ],
+            );
+        } else {
+            return response.body as CoordinatorServerCancellationResponse;
+        }
+    }
+
+    /**
+     * Batch version of softCancelOrderAsync. Requests multiple soft cancels
+     * @param   orders                An array of orders to cancel.
+     * @return  CoordinatorServerCancellationResponse. See [Cancellation Response](https://github.com/0xProject/0x-protocol-specification/blob/master/v2/coordinator-specification.md#response).
+     */
+    public async batchSoftCancelOrdersAsync(orders: SignedOrder[]): Promise<CoordinatorServerCancellationResponse[]> {
+        assert.doesConformToSchema('orders', orders, schemas.ordersSchema);
+        const makerAddress = getMakerAddressOrThrow(orders);
+        assert.isSenderAddressAsync('makerAddress', makerAddress, this._web3Wrapper);
+        const data = this._transactionEncoder.batchCancelOrdersTx(orders);
+
+        // create lookup tables to match server endpoints to orders
+        const feeRecipientsToOrders: { [feeRecipient: string]: SignedOrder[] } = {};
+        for (const order of orders) {
+            if (feeRecipientsToOrders[order.feeRecipientAddress] === undefined) {
+                feeRecipientsToOrders[order.feeRecipientAddress] = [] as SignedOrder[];
+            }
+            feeRecipientsToOrders[order.feeRecipientAddress].push(order);
+        }
+
+        const serverEndpointsToFeeRecipients: { [feeRecipient: string]: string[] } = {};
+        for (const feeRecipient of Object.keys(feeRecipientsToOrders)) {
+            const endpoint = await this._getServerEndpointOrThrowAsync(feeRecipient);
+            if (serverEndpointsToFeeRecipients[endpoint] === undefined) {
+                serverEndpointsToFeeRecipients[endpoint] = [];
+            }
+            serverEndpointsToFeeRecipients[endpoint].push(feeRecipient);
+        }
```

Comment:
This logic is duplicated in at least two locations. Let's functionally decompose and make it DRY.

Reply:
I simplified this part but left the rest separate (handling server responses) because slightly different things need to happen for fills vs cancels.